src/main.rs

```diff
@@ -16,19 +16,77 @@ mod snapshot_tests;
 fn main() {
     human_panic::setup_panic!();
 
-    let Cargo::SemverChecks(args) = Cargo::parse();
+    let Cargo::SemverChecks(mut args) = Cargo::parse();
+
+    let feature_flags = HashSet::from_iter(args.unstable_features);
 
     configure_color(args.color_choice);
     let mut config = GlobalConfig::new();
+    config.set_log_level(args.check_release.verbosity.log_level());
+    config.set_feature_flags(feature_flags);
+
+    validate_feature_flags(&mut config, &mut args.unstable_options);
+
+    if config.feature_flag_enabled(FeatureFlag::HELP) {
+        config
+            .log_info(|config| {
+                let header = Style::new()
+                    .bold()
+                    .fg_color(Some(Color::Ansi(AnsiColor::Cyan)));
+                let option = Style::new().bold();
+
+                writeln!(config.stderr(), "{header}Unstable options{header:#}\n")?;
+                writeln!(
+                    config.stderr(),
+                    "{header}{:<20}{header:#}Description",
+                    "-Z name",
+                )?;
+
+                for flag in FeatureFlag::ALL_FLAGS.iter().filter(|x| !x.stable) {
+                    write!(config.stderr(), "{option}{:<20}{option:#}", flag.id)?;
+
+                    if let Some((first, rest)) = flag.help.and_then(|x| x.split_first()) {
+                        writeln!(config.stderr(), "{first}")?;
 
-    if args.bugreport {
+                        for line in rest {
+                            writeln!(config.stderr(), "{:<20}{line}", "")?;
+                        }
+                    } else {
+                        writeln!(config.stderr())?;
+                    }
+                }
+
+                Ok(())
+            })
+            .expect("write failed");
+
+        std::process::exit(0);
+    } else if args.unstable_options.unstable_help {
+        /// Unstable command-line options.  Run `cargo semver-checks --help` to see stable help.
+        #[derive(Parser)]
+        #[clap(
+            disable_help_flag = true,
+            override_usage = "cargo semver-checks -Z unstable-options [OPTIONS] <stable-flags>",
+            mut_args = |arg| arg.hide(false),
+        )]
+        struct HelpPrinter {
+            #[command(flatten)]
+            args: UnstableOptions,
+        }
+
+        write!(
+            config.stderr(),
+            "{}",
+            HelpPrinter::command().render_long_help()
+        )
+        .expect("print failed");
+
+        std::process::exit(0);
+    } else if args.bugreport {
```

Comment:
Is there a way we can refactor this so that the `args.bugreport` branch happens as early as possible?

Two reasons:
- `--bugreport` should dominate all other flags. We don't want to skip reporting a bug by printing unstable features' help text.
- Every line of code we run before `args.bugreport` is another opportunity for a bug to prevent the bug report from happening. We want as few of those as possible.

Reply:
Now that I think about it, it's probably a good idea to have a test for `--bugreport` actually ... printing a report 😁 The exact system details in the report aren't relevant and can be masked out, we just care that the command has an effect and something vaguely of the right shape is printed.